File: CodeOfKutulu/src/test/scala/ExplorerTest.scala

class ExplorerTest
  extends UnitTest {
  "Explorer" should {
    "renvoyer le résultat attendu" when {
      "l'allié est à la même case" in {
        val givenLab = LabyrinthBuilder.givenLabFromFile("src/test/resources/depthSearch/emptyLab/given.txt", 6, 6)
        val givenExplorer1 = Explorer(0, Position(2, 2), 9999, 0, 0, givenLab)
        val givenExplorer2 = Explorer(1, Position(2, 2), 9999, 0, 0, givenLab)

        val friendRange: Option[Position] = givenExplorer1.goInFriendRange(Seq(givenExplorer2), givenLab)
        friendRange.get shouldEqual Position(2, 2)
      }

      "l'allié est à une case d'eccart" in {
        val givenLab = LabyrinthBuilder.givenLabFromFile("src/test/resources/depthSearch/emptyLab/given.txt", 6, 6)
        val givenExplorer1 = Explorer(0, Position(1, 1), 9999, 0, 0, givenLab)
        val givenExplorer2 = Explorer(1, Position(2, 2), 9999, 0, 0, givenLab)

        val friendRange: Option[Position] = givenExplorer1.goInFriendRange(Seq(givenExplorer2), givenLab)
        friendRange.get shouldEqual Position(1, 1)
      }

      "l'allié est ailleurs" in {
        val givenLab = LabyrinthBuilder.givenLabFromFile("src/test/resources/depthSearch/emptyLab/given.txt", 6, 6)
        val givenExplorer1 = Explorer(0, Position(1, 1), 9999, 0, 0, givenLab)
        val givenExplorer2 = Explorer(1, Position(4, 4), 9999, 0, 0, givenLab)

        val friendRange: Option[Position] = givenExplorer1.goInFriendRange(Seq(givenExplorer2), givenLab)
        friendRange.get shouldEqual Position(2, 4)
      }

      "l'allié ne peut être atteint" in {
        val givenLab = LabyrinthBuilder.givenLabFromFile("src/test/resources/depthSearch/simpleCross/given.txt", 5, 5)
        val givenExplorer1 = Explorer(0, Position(2, 2), 9999, 0, 0, givenLab)
        val givenExplorer2 = Explorer(1, Position(4, 4), 9999, 0, 0, givenLab)

        val friendRange: Option[Position] = givenExplorer1.goInFriendRange(Seq(givenExplorer2), givenLab)
        friendRange shouldBe None
      }
    }
  }
}

File: CodeOfKutulu/src/test/java/ExplorerTest.java
import org.junit.Test;

import java.util.Collections;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

public class ExplorerTest {

    @Test
    public void allie_memeCase() {
        Player.kutuluGrid = GridBuilder.kutuluGridFromFile("src/test/resources/depthSearch/emptyLab/given.txt", 6, 6);
        Explorer givenExplorer1 = new Explorer(0, new Location(2, 2), 9999, 0, 0);
        Explorer givenExplorer2 = new Explorer(1, new Location(2, 2), 9999, 0, 0);

        Location friendRange = givenExplorer1.goInFriendRange(Collections.singletonList(givenExplorer2));
        assertThat(friendRange).isEqualTo(new Location(2, 2));
    }

    @Test
    public void allie_uneCase() {
        Player.kutuluGrid = GridBuilder.kutuluGridFromFile("src/test/resources/depthSearch/emptyLab/given.txt", 6, 6);
        Explorer givenExplorer1 = new Explorer(0, new Location(1, 1), 9999, 0, 0);
        Explorer givenExplorer2 = new Explorer(1, new Location(2, 2), 9999, 0, 0);

        Location friendRange = givenExplorer1.goInFriendRange(Collections.singletonList(givenExplorer2));
        assertThat(friendRange).isEqualTo(new Location(1, 1));
    }

    @Test
    public void allie_loin() {
        Player.kutuluGrid = GridBuilder.kutuluGridFromFile("src/test/resources/depthSearch/emptyLab/given.txt", 6, 6);
        Explorer givenExplorer1 = new Explorer(0, new Location(1, 1), 9999, 0, 0);
        Explorer givenExplorer2 = new Explorer(1, new Location(4, 4), 9999, 0, 0);

        Location friendRange = givenExplorer1.goInFriendRange(Collections.singletonList(givenExplorer2));
        assertThat(friendRange).isEqualTo(new Location(2, 4));
    }

    @Test
    public void allie_nonAccessible() {
        Player.kutuluGrid = GridBuilder.kutuluGridFromFile("src/test/resources/depthSearch/simpleCross/given.txt", 5, 5);
        Explorer givenExplorer1 = new Explorer(0, new Location(2, 2), 9999, 0, 0);
        Explorer givenExplorer2 = new Explorer(1, new Location(4, 4), 9999, 0, 0);

        Location friendRange = givenExplorer1.goInFriendRange(Collections.singletonList(givenExplorer2));
        assertThat(friendRange).isNull();
    }
}
